File: native_reassembler_module/cpp/reassembler.cpp
#include "reassembler.h"
#include "packer.h"
#include "util.h"


Reassembler::Reassembler() : m_reassembled(false) {
}

void Reassembler::addRect(const Rect &rect) {
    if (m_reassembled)
        throw std::runtime_error("Cannot add more rectangles after reassembly.");
    m_rects.push_back(rect);
}

void Reassembler::reassemble(const cv::Mat &srcImg,
                             cv::Mat &dstImg,
                             int initial_width,
                             RectSortingMethod sorting_method,
                             bool autosize,
                             int border,
                             int margin,
                             ROIExtractor *roi_extractor) {
    if (m_reassembled)
        throw std::runtime_error("Already reassembled");

    Stopwatch sw_total_time;
    Stopwatch sw;

    // Step 0: Extract green regions
    if (roi_extractor != nullptr) {
        std::vector<Rect> green_rects = roi_extractor->extract_roi(srcImg);
        for (const Rect &rect: green_rects) {
            addRect(rect);
        }
    }
    sw.stop_and_print("extract roi");

    // Step 1: Finalise rectangles (add margins and borders)
    int imgw = srcImg.cols;
    int imgh = srcImg.rows;
    std::vector<Rect> expanded_rects;
    for (const Rect &rect: m_rects) {
        int x1 = rect.x;
        int y1 = rect.y;
        int w = rect.w;
        int h = rect.h;
        int x2 = x1 + w;
        int y2 = y1 + h;

        int extra = margin + border;
        x1 = std::max(x1 - extra, 0);
        y1 = std::max(y1 - extra, 0);
        x2 = std::min(x2 + extra, imgw);
        y2 = std::min(y2 + extra, imgh);
        expanded_rects.emplace_back(x1, y1, x2 - x1, y2 - y1);
        auto &&r = expanded_rects.back();
    }

    // Step 2: Sort rects and pack
    auto compare_func = getRectComparator(sorting_method);

    std::vector<Rect> sorted_rects = expanded_rects;
    std::sort(sorted_rects.begin(), sorted_rects.end(), compare_func);
    sw.stop_and_print("sort rects");

    std::vector<RectMapping> mappings;
    if (autosize) {
        ResizablePacker packer;
        mappings = packer.fit(sorted_rects);
    } else {
        Packer packer(initial_width, initial_width);
        mappings = packer.fit(sorted_rects);
    }
    this->m_mappings = mappings;
    sw.stop_and_print("fit");

    // Step 3: Construct the reassembled image
    cv::Mat result_img = draw_rects(srcImg, mappings, border);
    sw.stop_and_print("construct img");

    m_reassembled = true;

    // Collect stats
    int effective_area = 0;
    int n_fail = 0;
    for (const RectMapping &mapping: mappings) {
        if (mapping.dst) {
            effective_area += mapping.dst->area();
        } else {
            n_fail += 1;
        }
    }

    double total_time = sw_total_time.stop();

    int total_area = result_img.cols * result_img.rows;
    std::cout << "Result image size: " << result_img.cols << "x" << result_img.rows
              << "  effective_area=" << effective_area
              << "  total_area=" << total_area
              << "  utilisation=" << static_cast<double>(effective_area) / total_area
              << "  n_fail=" << n_fail
              << "  total_time=" << total_time << "ms:" << std::endl;

    dstImg = std::move(result_img);
}

cv::Mat Reassembler::draw_rects(const cv::Mat &srcImg,
                                const std::vector<RectMapping> &rect_mappings,
                                int border) {
    int canvas_width = 64;
    int canvas_height = 64;
    for (const RectMapping &rm: rect_mappings) {
        if (rm.dst) {
            canvas_width = std::max(canvas_width, rm.dst->x + rm.dst->w);
            canvas_height = std::max(canvas_height, rm.dst->y + rm.dst->h);
        }
    }

    cv::Mat canvas(canvas_height, canvas_width, srcImg.type(), cv::Scalar(0, 0, 0));

// TODO do we parallelize this?
#pragma omp parallel for
    for (int i = 0; i < rect_mappings.size(); i++) {
        const RectMapping &rm = rect_mappings[i];
        if (rm.dst) {
            int dst_x_start = rm.dst->x + border;
            int dst_y_start = rm.dst->y + border;
            int dst_x_end = rm.dst->x + rm.dst->w - border;
            int dst_y_end = rm.dst->y + rm.dst->h - border;

            int src_x_start = rm.src.x + border;
            int src_y_start = rm.src.y + border;
            int src_x_end = rm.src.x + rm.src.w - border;
            int src_y_end = rm.src.y + rm.src.h - border;

            cv::Rect dst_roi(dst_x_start, dst_y_start, dst_x_end - dst_x_start, dst_y_end - dst_y_start);
            cv::Rect src_roi(src_x_start, src_y_start, src_x_end - src_x_start, src_y_end - src_y_start);

            if (dst_roi.width > 0 && dst_roi.height > 0 && src_roi.width > 0 && src_roi.height > 0 &&
                src_roi.x >= 0 && src_roi.y >= 0 && src_roi.x + src_roi.width <= srcImg.cols &&
                src_roi.y + src_roi.height <= srcImg.rows &&
                dst_roi.x >= 0 && dst_roi.y >= 0 && dst_roi.x + dst_roi.width <= canvas.cols &&
                dst_roi.y + dst_roi.height <= canvas.rows) {
                srcImg(src_roi).copyTo(canvas(dst_roi));
            }
        }
    }

    return canvas;
}

std::vector<RectMapping> Reassembler::reverse_mapping(const std::vector<Rect> &rects) const {
    if (!m_reassembled)
        throw std::runtime_error("Not reassembled");

    std::vector<RectMapping> result = map0(rects, rects, rects, true);
    return result;
}

std::vector<RectMapping> Reassembler::mapping(const std::vector<Rect> &rects) const {
    if (!m_reassembled)
        throw std::runtime_error("Not reassembled");

    std::vector<RectMapping> result = map0(rects, rects, rects, false);
    return result;
}

std::vector<RectMapping> Reassembler::map0(const std::vector<Rect> &rect_mapped_space,
                                           const std::vector<Rect> &rect_reference,
                                           const std::vector<Rect> &rects,
                                           bool rev) const {
    std::vector<RectMapping> result;
    std::vector<std::vector<double>> ious = box_ioa1(rect_mapped_space, rect_reference);

    for (size_t i = 0; i < rect_mapped_space.size(); ++i) {
        const Rect &rect = rects[i];
        size_t ci = std::distance(ious[i].begin(), std::max_element(ious[i].begin(), ious[i].end()));
        double iou = ious[i][ci];

        for (size_t j = 0; j < m_mappings.size(); ++j) {
            const RectMapping &rm = m_mappings[j];
            if ((rev && rm.dst->contains_rect(rect)) || (!rev && rm.src.contains_rect(rect))) {
                iou = 1.0;
                ci = j;
                break;
            }
        }
        RectMapping candidate = m_mappings[ci];

        if (!candidate.dst || iou < 0.6) {
            result.emplace_back(rect, std::nullopt);
        } else {
            const Rect &dst = rev ? candidate.dst.value() : candidate.src;
            const Rect &src = rev ? candidate.src : candidate.dst.value();
            int dx = rect.x - dst.x;
            int dy = rect.y - dst.y;
            Rect mapped_rect(src.x + dx, src.y + dy, rect.w, rect.h);
            result.emplace_back(rect, mapped_rect);
        }
    }

    return result;
}

std::vector<RectMapping> Reassembler::get_raw_mappings() const {
    return m_mappings;
}

File: roiyolowd/reassembler.py
import time
from typing import List, Optional, Tuple

import numpy as np

from roiyolowd.roi_extractor import ROIExtractor, ExGIGreenExtractor
from roiyolowd.util import box_ioa1, Rect, RectMapping, Stopwatch


# Reference:
# https://codeincomplete.com/articles/bin-packing/demo/

class RectSorting:
    """
    A few comparators for sorting Rect
    """
    WIDTH_DESC = lambda rect: rect.w
    HEIGHT_DESC = lambda rect: rect.h
    AREA_DESC = lambda rect: rect.w * rect.h
    MAXSIDE_DESC = lambda rect: max(rect.w, rect.h)


class Block:
    def __init__(self, x: int, y: int, w: int, h: int):
        self.x = x
        self.y = y
        self.w = w
        self.h = h
        self.used = False
        self.down = None
        self.right = None


class Packer:
    def __init__(self, w: int, h: int):
        self.root: Block = Block(0, 0, w, h)

    def fit(self, rects: List[Rect]):
        result: List[RectMapping] = []
        for rect in rects:
            node = self.find_node(self.root, rect.w, rect.h)
            if node:
                fit = self.split_node(node, rect.w, rect.h)
                fit_rect = Rect(fit.x, fit.y, rect.w, rect.h)
                mapping = RectMapping(rect, fit_rect)
                result.append(mapping)
            else:
                mapping = RectMapping(rect, None)
                result.append(mapping)
        return result

    def find_node(self, root: Block, w: int, h: int) -> Optional[Block]:
        if root.used:
            return self.find_node(root.right, w, h) or self.find_node(root.down, w, h)
        elif w <= root.w and h <= root.h:
            return root
        else:
            return None

    def split_node(self, node: Block, w: int, h: int) -> Block:
        node.used = True
        node.down = Block(node.x, node.y + h, node.w, node.h - h)
        node.right = Block(node.x + w, node.y, node.w - w, h)
        return node


class ResizablePacker:
    def __init__(self):
        self.root: Optional[Block] = None

    def fit(self, rects: List[Rect]):
        if not rects or len(rects) == 0:
            return []

        w = rects[0].w
        h = rects[0].h
        self.root = Block(0, 0, w, h)
        result: List[RectMapping] = []

        for rect in rects:
            node = self.find_node(self.root, rect.w, rect.h)
            fit_rect: Optional[Rect] = None
            if node:
                fit = self.split_node(node, rect.w, rect.h)
                fit_rect = Rect(fit.x, fit.y, rect.w, rect.h)
            else:
                fit = self.grow_node(rect.w, rect.h)
                if fit:
                    fit_rect = Rect(fit.x, fit.y, rect.w, rect.h)
            mapping = RectMapping(rect, fit_rect)
            result.append(mapping)
        return result

    def find_node(self, root: Block, w: int, h: int) -> Optional[Block]:
        if root.used:
            return self.find_node(root.right, w, h) or self.find_node(root.down, w, h)
        elif w <= root.w and h <= root.h:
            return root
        else:
            return None

    def split_node(self, node: Block, w: int, h: int) -> Block:
        node.used = True
        node.down = Block(node.x, node.y + h, node.w, node.h - h)
        node.right = Block(node.x + w, node.y, node.w - w, h)
        return node

    def grow_node(self, w: int, h: int) -> Optional[Block]:
        can_grow_down = w <= self.root.w
        can_grow_right = h <= self.root.h

        should_grow_right = can_grow_right and (self.root.h >= self.root.w + w)
        should_grow_down = can_grow_down and (self.root.w >= self.root.h + h)

        if should_grow_right:
            return self.grow_right(w, h)
        elif should_grow_down:
            return self.grow_down(w, h)
        elif can_grow_right:
            return self.grow_right(w, h)
        elif can_grow_down:
            return self.grow_down(w, h)
        else:
            return None

    def grow_right(self, w: int, h: int) -> Optional[Block]:
        root_old = self.root
        self.root = Block(0, 0, root_old.w + w, root_old.h)
        self.root.used = True
        self.root.down = root_old
        self.root.right = Block(root_old.w, 0, w, root_old.h)
        node = self.find_node(self.root, w, h)
        if node:
            return self.split_node(node, w, h)
        else:
            return None

    def grow_down(self, w: int, h: int) -> Optional[Block]:
        root_old = self.root
        self.root = Block(0, 0, root_old.w, root_old.h + h)
        self.root.used = True
        self.root.down = Block(0, root_old.h, root_old.w, h)
        self.root.right = root_old
        node = self.find_node(self.root, w, h)
        if node:
            return self.split_node(node, w, h)
        else:
            return None


class Reassembler:
    """
    Usage:
    r = Reassembler()
    r.addRect(...) // optional
    reassembled_img = r.reassemble(img)

    """
    __default_roi_extractor_sentinel = object()

    def __init__(self):
        self.rects: List[Rect] = []
        self.mappings: List[RectMapping] = []
        self.reassembled = False

    def addRect(self, rect: Rect) -> None:
        if self.reassembled:
            raise Exception("Cannot add more rectangles after reassembly.")
        self.rects.append(rect)

    def reassemble(self, srcImg: np.ndarray,
                   packer_width: int = 640,
                   sorting_method: RectSorting = RectSorting.HEIGHT_DESC,
                   use_resizable_packer: bool = True,
                   border_thickness: int = 3,
                   padding_size: int | Tuple[int, int] = 8,
                   roi_extractor: Optional[ROIExtractor] = __default_roi_extractor_sentinel) -> np.ndarray:
        """
        Args:
            srcImg: The input image from which to extract ROIs and reassemble them.
            packer_width: The width for the fixed packer. A reasonable value is the square root of the (estimated) total area of all ROIs.
            If using a resizable packer (`use_resizable_packer=True`), this can be ignored or set to an arbitrary value.
            sorting_method: An appropriate ROI sorting method is required to achieve optimal space utilisation. The default one (HEIGHT_DESC) is good enough for both fixed size and resizable packer.
            use_resizable_packer: Frame size are automatically adjusted during fitting. packer_width does not take effect when use_resizable_packer is True.
            border_thickness: The thickness (in pixels) of the black borders added to all sides of each ROI.
            padding_size: The number of pixels by which to expand each side of the rectangles outward from their center.
            This enlarges the cropped area uniformly while keeping it centered around its original position.
            If a 2-tuple is provided, the padding size is randomized within the specified lower and upper bounds (the 2 numbers in the tuple).
            roi_extractor:

        Returns:
            The reassembled image.

        """
        if self.reassembled:
            raise Exception("Already reassembled")

        sw = Stopwatch()
        # Step 0: Extract green regions
        if roi_extractor is not None:
            if roi_extractor == self.__default_roi_extractor_sentinel:
                roi_extractor = ExGIGreenExtractor()
            green_rects: List[Rect] = roi_extractor.extract_roi(srcImg)
            for rect in green_rects:
                self.addRect(rect)
        sw.stop("extract green")

        # Step 1: Finalise rectangles (add margins and borders)
        # This place is a bit messy, but I don't think it's worth a refactor, because
        # it's still a relatively simple task.
        imgh, imgw = srcImg.shape[:2]
        expanded_rects: List[Rect] = []
        if isinstance(padding_size, int):  # Constant size margin
            extra = padding_size + border_thickness
            for rect in self.rects:
                x1, y1, w, h = rect.x, rect.y, rect.w, rect.h
                x2, y2 = x1 + w, y1 + h
                x1, y1 = max(x1 - extra, 0), max(y1 - extra, 0)
                x2, y2 = min(x2 + extra, imgw), min(y2 + extra, imgh)
                expanded_rects.append(Rect(x1, y1, x2 - x1, y2 - y1))
                rect = expanded_rects[len(expanded_rects) - 1]
        else:  # Margin with size ranging from the given range
            margin_lo, margin_hi = padding_size
            for rect in self.rects:
                x1, y1, w, h = rect.x, rect.y, rect.w, rect.h
                x2, y2 = x1 + w, y1 + h
                mL, mR, mT, mB = np.random.randint(margin_lo, margin_hi, size=4)
                x1, y1 = max(x1 - mL, 0), max(y1 - mT, 0)
                x2, y2 = min(x2 + mR, imgw), min(y2 + mB, imgh)
                expanded_rects.append(Rect(x1, y1, x2 - x1, y2 - y1))
        sw.stop("margins")

        # Step 2: sort rects and pack
        start_t = time.time()
        sorted_rects = sorted(expanded_rects, key=sorting_method, reverse=True)
        if use_resizable_packer:
            packer = ResizablePacker()
        else:
            packer = Packer(packer_width, packer_width)
        mappings = packer.fit(sorted_rects)
        self.mappings = mappings
        fit_el = int((time.time() - start_t) * 1000)
        sw.stop("fit")

        # Step 3: construct the reassembled image
        start_t = time.time()
        result_img = self.__draw_rects(srcImg, mappings, border_thickness)
        draw_el = int((time.time() - start_t) * 1000)
        sw.stop("img")

        # Collect stats
        effective_area = np.sum([i.dst.w * i.dst.h if i.dst else 0 for i in mappings])
        n_fail = np.sum([0 if i.dst else 1 for i in mappings])
        total_area = result_img.shape[0] * result_img.shape[1]
        print(f"{result_img.shape}   areaR={total_area / 640 / 640}   utilisation={effective_area / total_area}"
              f"  n_fail={n_fail}   fit_time={fit_el}ms   draw_time={draw_el}ms")

        self.reassembled = True
        return result_img

    def reverse_map(self, rects: List[Rect]) -> List[RectMapping]:
        """
        Map a list of rects from mapped image space back to original image space
        Args:
            rects:

        Returns:

        """
        if not self.reassembled:
            raise Exception("Not reassembled")

        rect_mapped_space = [(r.x, r.y, r.x + r.w, r.y + r.h) for r in rects]
        rect_reference = [(r.dst.x, r.dst.y, r.dst.x + r.dst.w, r.dst.y + r.dst.h) for r in self.mappings]

        result: List[RectMapping] = self.__map0(rect_mapped_space, rect_reference, rects, rev=True)
        return result

    def forward_map(self, rects: List[Rect]) -> List[RectMapping]:
        """
        Map a list of rects from original image space to reassembled image space
        Args:
            rects:

        Returns:

        """
        if not self.reassembled:
            raise Exception("Not reassembled")

        rect_mapped_space = [(r.x, r.y, r.x + r.w, r.y + r.h) for r in rects]
        rect_reference = [(r.src.x, r.src.y, r.src.x + r.src.w, r.src.y + r.src.h) for r in self.mappings]

        result: List[RectMapping] = self.__map0(rect_mapped_space, rect_reference, rects, rev=False)
        return result

    def __draw_rects(self, srcImg: np.ndarray,
                     rect_mappings: List[RectMapping],
                     border: int = 0) -> np.ndarray:
        """
        Construct a reassembled image using rect_mappings.
        Args:
            srcImg:
            rect_mappings:
            border:

        Returns:

        """
        canvas_width = max([rm.dst.x + rm.dst.w for rm in rect_mappings if rm is not None] + [64])
        canvas_height = max([rm.dst.y + rm.dst.h for rm in rect_mappings if rm is not None] + [64])
        canvas = np.full((canvas_height, canvas_width, 3), 0, dtype=np.uint8)

        for rect_map in rect_mappings:
            if rect_map.dst:
                dst_x_start = rect_map.dst.x + border
                dst_x_end = rect_map.dst.x + rect_map.dst.w - border
                dst_y_start = rect_map.dst.y + border
                dst_y_end = rect_map.dst.y + rect_map.dst.h - border

                src_x_start = rect_map.src.x + border
                src_x_end = rect_map.src.x + rect_map.src.w - border
                src_y_start = rect_map.src.y + border
                src_y_end = rect_map.src.y + rect_map.src.h - border

                canvas[dst_y_start:dst_y_end, dst_x_start:dst_x_end] = srcImg[src_y_start:src_y_end, src_x_start:src_x_end]

        return canvas

    def __map0(self, rect_mapped_space: List[Tuple], rect_reference: List[Tuple], rects: List[Rect], rev: bool):
        result: List[RectMapping] = []
        ious = box_ioa1(np.array(rect_mapped_space), np.array(rect_reference))
        for i in range(0, len(rect_mapped_space)):
            rect = rects[i]
            ci = np.argmax(ious[i])
            candidate = self.mappings[ci]
            iou = ious[i][ci]

            # Choose the ROI that has the most overlapping with the rectangle, and use that ROI as the reference point to
            # map the rectangle location back to the original image.
            for j, rm in enumerate(self.mappings):
                if (rev and rm.dst.contains_rect(rect)) or (not rev and rm.src.contains_rect(rect)):
                    candidate = rm
                    iou = 1
                    break

            if not candidate or iou < 0.6:
                result.append(RectMapping(rect, None))
            else:
                dst = candidate.dst if rev else candidate.src
                src = candidate.src if rev else candidate.dst
                dx = rect.x - dst.x
                dy = rect.y - dst.y
                mapped_rect = Rect(src.x + dx, src.y + dy, rect.w, rect.h)
                result.append(RectMapping(rect, mapped_rect))

        return result


def create_reassembler(use_native: bool = False) -> Reassembler:
    if use_native:
        from roiyolowd.native_reassembler import NativeReassembler
        return NativeReassembler()
    else:
        return Reassembler()
